bin/ch/ch.cpp

```diff
@@ -196,48 +207,536 @@ static void CALLBACK PromiseContinuationCallback(JsValueRef task, void *callback
     MessageQueue * messageQueue = (MessageQueue *)callbackState;
 
     WScriptJsrt::CallbackMessage *msg = new WScriptJsrt::CallbackMessage(0, task);
+
+#if ENABLE_TTD
+    ChakraRTInterface::JsTTDNotifyHostCallbackCreatedOrCanceled(true, false, false, task, msg->GetId());
+#endif
+
     messageQueue->InsertSorted(msg);
 }
 
-HRESULT RunScript(const char* fileName, LPCWSTR fileContents, BYTE *bcBuffer, char *fullPath)
+void StartupDebuggerAsNeeded()
 {
-    HRESULT hr = S_OK;
-    MessageQueue * messageQueue = new MessageQueue();
-    WScriptJsrt::AddMessageQueue(messageQueue);
-    LPWSTR fullPathWide = nullptr;
+    if(dbgIPAddr == nullptr)
+    {
+        if(doTTDebug)
+        {
+            //we need to force the script context into dbg mode for replay even if we don't attach the debugger -- so do that here
+            ChakraRTInterface::JsTTDSetDebuggerForReplay();
+        }
+    }
+    else
+    {
+        //
+        //TODO: not sure how to set IP and other stuff here !!!
+        //
+        AssertMsg(false, "Broken");
 
-    IfJsErrorFailLog(ChakraRTInterface::JsSetPromiseContinuationCallback(PromiseContinuationCallback, (void*)messageQueue));
+        Debugger* debugger = Debugger::GetDebugger(chRuntime);
+        debugger->StartDebugging(chRuntime);
+    }
+}
+
+void CreateDirectoryIfNeeded(const char16* path)
+{
+#ifndef _WIN32
+    AssertMsg(false, "Not XPLAT yet.");
+#else
+    bool isPathDirName = (path[wcslen(path) - 1] == _u('\\'));
 
-    Assert(fileContents != nullptr || bcBuffer != nullptr);
-    // TODO: Remove this code in a future iteration once Utf8 versions of the Jsrt API is implemented
-    IfFailGo(Helpers::NarrowStringToWideDynamic(fullPath, &fullPathWide));
+    size_t fplength = (wcslen(path) + 2);
+    char16* fullpath = new char16[fplength];
+    fullpath[0] = _u('\0');
 
-    JsErrorCode runScript;
-    if (bcBuffer != nullptr)
+    wcscat_s(fullpath, fplength, path);
+    if(!isPathDirName)
     {
-        runScript = ChakraRTInterface::JsRunSerializedScript(fileContents, bcBuffer, WScriptJsrt::GetNextSourceContext(), fullPathWide, nullptr /*result*/);
+        wcscat_s(fullpath, fplength, _u("\\"));
+    }
+
+    DWORD dwAttrib = GetFileAttributes(fullpath);
+    if((dwAttrib != INVALID_FILE_ATTRIBUTES) && (dwAttrib & FILE_ATTRIBUTE_DIRECTORY))
+    {
+        delete[] fullpath;
+        return;
+    }
+
+    BOOL success = CreateDirectory(fullpath, NULL);
+    if(!success)
+    {
+        DWORD lastError = GetLastError();
+        LPTSTR pTemp = NULL;
+        FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ARGUMENT_ARRAY, NULL, lastError, 0, (LPTSTR)&pTemp, 0, NULL);
+        fwprintf(stderr, _u(": %s"), pTemp);
```

Comment:
Other failure messages in ch go to stderr (e.g. GenerateLibraryByteCodeHeader message). Not sure what the preferred style is.